### ISSUE TITLE
Can't wait longer than 2147483.647 seconds

### ISSUE THREAD

**Issue by ericrosenbaum** [LLK/scratch-vm#1010]
### Expected Behavior

Any number of seconds entered into e.g. a "say" or "wait" block should cause the block to wait that number of seconds.

### Actual Behavior

Numbers of seconds over a certain large threshold cause the wait time to be zero. We often see kids typing in very large numbers like this.

With help from @rschamp, we found the exact threshold, which is 2147483.647 (1/1000 of max int, presumably because it is converted to milliseconds). So, waiting for 2147483.647 seconds seems to work (though we didn't actually wait for 24.8 days to find out), but waiting for 2147483.648 seconds waits for 0 instead:

![big-wait-bug](https://user-images.githubusercontent.com/567844/38202347-63575bf6-3669-11e8-80d6-6d601bc8e6e9.gif) 

### Steps to Reproduce

See gif

**Comment by apple502j** [comment]
In some programming languages, the max value of integers is 2147483647.
Inside the code, the block calculates 1000 times the seconds,
so 2147483.647 * 1000 = 2147483647
2147483.648 * 1000 = -2147483648

**Comment by thisandagain** [comment]
Adding 'help-wanted' to this. I would recommend modifying the math utilities to include a call to `Number.isSafeInteger`. If it fails, the result should be `253 - 1` (maximum double precision number as per IEEE-754).

**Comment by apple502j** [comment]
Or use unsigned int for seconds?

**Comment by Kenny2github** [comment]
@apple502j unsigned int just means the limit is higher, not gone. I would recommend an equivalent to Python 2's `long` integer type, if such a thing exists in JS.

**Comment by kchadha** [comment]
Sort of related, I noticed that we use two different number fields (and therefore different validation as to what characters can be typed in) for the `wait (n) seconds` block vs. the `say/think [hello] for (n) seconds` blocks. Should they be the same? The wait block does not allow typing in negative numbers  while the say/think blocks do.

**Comment by thisandagain** [comment]
@kchadha Good catch! That should be fixed. The say/think and glide blocks should use the same validation as the wait block.

**Comment by Kenny2github** [comment]
@kchadha but do the number inputs in the say/think blocks have this problem as well?

**Comment by kchadha** [comment]
@Kenny2github yeah, the say/think blocks are the ones mentioned in the original bug report above.

**Comment by towerofnix** [comment]
(This is tagged with "help-wanted", not "help wanted"; probably a mistake since the vast majority of other issues are tagged "help wanted", and GitHub only [recognizes](https://i.cubeupload.com/IscIiP.png) issues labeled "help wanted".)

**Comment by thisandagain** [comment]
@towerofnix Thanks for catching that. The `help-wanted` label is a legacy from before Github added it to all repos.

**Comment by apple502j** [comment]
Ken said that setTimeout can only accept numbers less than 2147483647msec.

**Comment by mrjacobbloom** [comment]
Is it really worth it to respect wait times that high? For any wait time more than a couple days you can probably never execute the rest of the script without much fear of incompatibility. The only kind of project I can imagine actually depending on egregiously long wait times is some kind of cloud server -- or a jump scare someone plans to leave running on someone else's computer

**Comment by mzgoddard** [comment]
The blocks probably shouldn't use setTimeout. Since the outcome of the waiting only starts in a future step of the scratch engine, I think the blocks should count the frame equivalent of the amount of time they wait. Most of that is in response to this issue: https://github.com/LLK/scratch-vm/issues/904 and my comment on it https://github.com/LLK/scratch-vm/issues/904#issuecomment-362680457.

This could be done by setting a future time on the execution context (like how loops count) and yielding until that time is reached. Or setting the time to wait and decreases it by runtime.currentStepTime until its less than or equal to 0.

**Comment by AmazingMech2418** [comment]
I have an idea! Instead of setTimeout, you should use a setInterval that changes by 0.1 every 0.1 seconds or use the current timer variable for the timer block and only start the next block when the timer reaches the inputted number of seconds.

**Comment by AmazingMech2418** [comment]
I just added a PR. I don't know all of the locations with the other blocks with a setTimeout, so if you know of any, please let me know so I can update my PR.

**Comment by AmazingMech2418** [comment]
The errors are all node modules that aren't present since they can only be found in gh-pages and not develop.

**Comment by ericrosenbaum** [comment]
Thanks for your thoughts on this folks! The issue here is the integer overflow problem, so any PR addressing this issue should be a fix for that. We should go with the fix recommended by @thisandagain above.

Sorry @amazinigmech2418, we will need to close your PR. It seems to be addressing the one @mzgoddard mentioned, which is actually documented in https://github.com/LLK/scratch-vm/issues/904, but is not the fix for that issue which we want. For that one, we should be using the VM's stack timer and yield mechanisms (as I did for the timing in the music extension, e.g. [here](https://github.com/LLK/scratch-vm/blob/8739a524eb1241590098b00f003df5f234207fff/src/extensions/scratch3_music/index.js#L638-L657)).

**Comment by joker314** [comment]
Note that the numbers for which Scratch breaks are actually safe integers (i.e., `Number.isSafeInteger(2147483648)` is `true`.

This is because that number is `2 ** 31`, but `Number.MAX_SAFE_INTEGER` is `2 ** 53 - 1`.

This is a problem that should only affect `setTimeout`, because it [apparently](https://stackoverflow.com/questions/3468607/why-does-settimeout-break-for-large-millisecond-delay-values) uses a signed 32-bit integer internally.

We could create something as a maths utility to convert numbers to signed 32-bit numbers (e.g. `Math.max(n, 2 ** 31 - 1)`. We could also create a custom `setTimeout` function (e.g. `setSafeTimeout`). I prefer the former.

**Comment by ericrosenbaum** [comment]
@thisandagain discussed this and we decided to replace setTimeout in each case with a system that uses the stack timer and yielding. We'll use `Number.isSafeInteger` to prevent the specific problem described in this issue.
